BUM SidePanel: Implement Enroll (learners) to classes functionality

LianaHarris360 opened the pull request:
<!--
 1. Following guidance below, replace …'s with your own words
 2. After saving the PR, tick of completed checklist items
 3. Skip checklist items that are not applicable or not necessary
 4. Delete instruction/comment blocks
-->

## Summary
<!--
 * description of the change
 * manual verification steps performed
 * screenshots if the PR affects the UI
-->
This pull request updates the EnrollLearnersSidePanel component to carry out the bulk action to enroll learners into one or more classes.

Functionality added:
- A selectable and filterable list of unchecked classes are shown when the EnrollLearnersSidePanel component is opened.
-  A loading spinner is shown within the side panel while the API request is being made and the side panel closes once the request has succeeded.
- A notice is shown at the top of the side panel when an API request fails.
- If the Admin has not selected a class, the **Cancel** button closes the side panel without confirmation. Otherwise, a confirmation modal appears.
- After the selected users are enrolled, the side panel closes and a `KModal` allowing the enrollment to be undone is displayed.
-`MembershipResource` is used to bulk enroll or un-enroll (undo) learners.
![enrollSidePanelError](https://github.com/user-attachments/assets/c9ce19f1-6ae5-4c18-a79f-7a7a2782a4f6)

Enrolling users using the side panel:

https://github.com/user-attachments/assets/aec4216b-996d-4460-b5fd-24fa85589843


Enrolling and undoing enrollment using the side panel:

https://github.com/user-attachments/assets/056abaea-d637-4675-a72a-7c563d96ca9e




## References
<!--
 * references to related issues and PRs
 * links to mockups or specs for new features
 * links to the diffs for any dependency updates, e.g. in iceqube or the perseus plugin
-->
Closes #13406 

## Reviewer guidance
<!--
 * how can a reviewer test these changes?
 * are there any risky areas that deserve extra testing
-->
1. On the Facility > Users page, select one or most users and click the enroll icon button.
2. Once the side panel opens, confirm that the correct warnings apply based on if the selected users are coaches, learners, and not enrolled in any classes.
3. Select a class to enroll the users into and undo the enrollment to confirm that the functionality works as intended. **Note:** Users that are already enrolled into the selected class(es) will not be affected by selecting ENROLL or UNDO.

LianaHarris360 committed:
Add enroll side panel bulk action strings

LianaHarris360 committed:
Add bulk enroll learners to classes functionality to side panel

LianaHarris360 committed:
Wrap buttons in KButtonGroup

nucleogenesis commented on kolibri/plugins/facility/assets/src/views/UserPage/SidePanels/EnrollLearnersSidePanel.vue:
```diff
@@ -1,40 +1,358 @@
+        memberships: [],
+        enrollments: [],
+      };
+    },
```
nitpick: To keep data definitions all in the same place these could be defined as refs in the setup function

nucleogenesis replied:
The same nitpick goes for the computed properties could be defined w/ `computed()` and the methods could be plain functions in setup function.

Not a blocker

nucleogenesis commented on kolibri/plugins/facility/assets/src/views/UserPage/SidePanels/EnrollLearnersSidePanel.vue:
```diff
@@ -1,40 +1,358 @@
+      @submit="handleUndoEnrollments"
+    >
+      <KCircularLoader v-if="loading" />
+      <span v-else>{{
```
Idk if we need a bit of line-height on this or a height to the span maybe... or maybe `overflow-y: visible` or something to get rid of the little baby scroll bar.

AlexVelezLl replied:
I also had this issue, and I fixed this by adjusting the line-height of the span containing the text https://github.com/learningequality/kolibri/pull/13538/files#diff-8f1c72bbeb798d7eb4ddfc0c5d0c10983d8fc5b76711e6f9c4fb0aafb35e3507R111

nucleogenesis reviewed (commented):
One minor blocking comment & a couple non-blocking nitpicks in the code.

I tested things out locally and the whole workflow worked as expected! I'll mark it ready for @radinamatic to check it out!

pcenov commented:
Hi @LianaHarris360,
Here are my observations and notes, some of which I'm just mentioning as a first impression and could be considered as potential follow-ups or just general feedback not related to the changes made here:

1. It seems that it should be further discussed whether the icons should be enabled when there are no users selected. For example if I have not selected a user and I click the "Enroll to class" icon I am just seeing a "You've selected 0 users" message and the "Enroll" button is disabled so there's nothing that I can actually do but go back, which brings me to my initial state. Having the icons disabled will be consistent with similar icons at Coach > Quiz > New:

https://github.com/user-attachments/assets/48273542-eb9e-4927-b0ab-c175c414801f

2. In continuation to the first point it's not immediately clear that the first icon with the back arrow is the "Assign coaches" button. The tooltip clears things out but to me this will always be a "Go back" icon.

3. After enrolling a user or several users to multiple classes then when I click the "Undo" button I am seeing an error screen with a 503 error "Database is not available for write operations":

https://github.com/user-attachments/assets/7351b4c9-81d6-4c16-a431-8af102fd2aa9

4. Attempting to enroll an already deleted learner results in a success message:

https://github.com/user-attachments/assets/621d9ff0-a52c-4be5-9f5c-ced728c8a4d3

5. Working on smaller screens with the Users table is quite difficult as it involves a lot of scrolling up and down and trying to find the appropriate options and controls - this can be seen in the video above.

6. The classes are not sorted in anyway which makes finding the need class very difficult (forced to search for it):

https://github.com/user-attachments/assets/a851b79b-f893-4cd7-bca7-b8cd4cbf3004

7. When enrolling users to a class, there are 4 different messages taking a significant portion of the screen, 2 of which are with big yellow exclamation marks suggesting that I need to take immediate action while this is not necessarily the case. The last one is with just an info icon, but it's not clear why it's there and why any users would be affected. Also in the video below you can see that I am enrolling a super admin and an admin user to the class, but I'm notified only about the coach? Perhaps the icons can be removed and the text shortened to something like: You've selected NN users of which NN are not enrolled in any class yet. NN of the selected users are not learners. If nothing else then it should be further discussed whether the admin users should be listed along with the coach in the warning message. 

https://github.com/user-attachments/assets/2d72669b-4340-4eb1-982f-481159bec54b

8. The confirmation modal is also confusing. Firstly, when adding a single user the text should be changed to " 1 user has been enrolled". Also since I am intentionally enrolling users to classes, why would the main course of events be to press the blue UNDO button? In this case I would expect the button to the right to be called CONTINUE and the one to the left UNDO. The "Undo this?" part of the bold text should also be removed.

<img width="825" height="779" alt="confusing modal" src="https://github.com/user-attachments/assets/2ff68da2-9631-4a58-ba42-5b878f38aaf6" />

nucleogenesis commented:
@pcenov thanks for the review here! 

@LianaHarris360 I think 3 and 4 should be addressed and handled here. For the "enrolling a deleted user" edge case - we ought to expect some kind of error to come back during the process but if we're not getting one back we should talk about how to address that.

For the "database locked" error - maybe the Undo-ing code needs the same treatment you gave the Do-ing code when you ran into that issue before?

LianaHarris360 commented:
I agree, the other points are definitely ones we should mention to Tomiwa. For the enrolling a deleted learner, maybe before the requests are sent there should be a check to ensure that all of the learners being enrolled are still current `FacilityUsers` and not enroll the user ids that cannot be found.

AlexVelezLl commented on kolibri/plugins/facility/assets/src/views/UserPage/SidePanels/EnrollLearnersSidePanel.vue:
```diff
@@ -1,40 +1,358 @@
+      },
+      usersNotEnrolled() {
+        const enrolledUsers = new Set(this.classLearners);
+        return [...this.selectedUsers].filter(userId => !enrolledUsers.has(userId)).length;
```
(Nitpick) Curious to know if there was a reason to do the `[...this.selectedUsers]` array clone to do the filter? The array filter does not modify the original array, and I dont see any code that can modify the original array, so probably doing just this.selectedUsers.filter... would have the same effect?

LianaHarris360 replied:
`selectedUsers` is a set so it needs to be cloned into an array to be filtered. To your point about passing user objects to the `selectedUsers` to avoid refetching the same data, this will probably be changed at some point.

AlexVelezLl commented on kolibri/plugins/facility/assets/src/views/UserPage/SidePanels/EnrollLearnersSidePanel.vue:
```diff
@@ -1,40 +1,358 @@
+          getParams: { user_ids: { user_ids: Array.from(this.selectedUsers).join(',') } },
+        });
+        const userModelsPromise = Promise.all(
+          Array.from(this.selectedUsers).map(id => FacilityUserResource.fetchModel({ id })),
```
We have the `by_ids` filter in the FacilityUserViewset so we can load all users data in just one fetch by filtering by this param.

The other thing I was thinking about is that if we need the users actual data, we should be probably storing the user object in the selectedUsers array rather than just the id, so we can make operations with those objects later. This is how we manage the selected resources in the resource selection side panels, by storing the whole objejct instead of just the ids. This way we won't need to make any other fetch request to get this data.

LianaHarris360 replied:
I thought about this as well, I agree that it is redundant to refetch the users data, and for this specific usage, only the role is really needed but I would agree that passing more than just the user_ids would be ideal. Because of conflicting work that is currently ongoing on the `UserPage` component for the user selections, I assume this is something that should go into a follow-up refactoring issue @nucleogenesis

AlexVelezLl commented on kolibri/plugins/facility/assets/src/views/UserPage/SidePanels/EnrollLearnersSidePanel.vue:
```diff
@@ -1,40 +1,358 @@
+            return;
+          }
+          for (const membership of this.enrollments) {
+            await MembershipResource.saveCollection({
```
We can save all these enrollments in just one request, rather than iterating per class.

There is an error in the original code because we dont actually need to set the collection in the get params. When we make this request, we have nothing in the get params:
<img width="846" height="922" alt="Image" src="https://github.com/user-attachments/assets/dd89cde1-3cfb-4773-983c-8e7c58d4a37e" />

So there is no need to make multiple request for each collection, we can include all these objects from different collections in a single array in the `data` property.

AlexVelezLl commented on kolibri/plugins/facility/assets/src/views/UserPage/SidePanels/EnrollLearnersSidePanel.vue:
```diff
@@ -1,40 +1,358 @@
+          if (this.enrollments.length > 0) {
+            await Promise.all(
+              this.enrollments.flatMap(({ collection_id, user_ids }) =>
+                MembershipResource.deleteCollection({
```
Since we are already using the BulkDeleteMixin in the MembershipViewset https://github.com/lianaharris360/kolibri/blob/e14e30f4e5549f8ece836c4a7f20b3054b0ce4e6/kolibri/core/auth/api.py#L649, we can also add a `by_ids` filter here, so we will be able to filter by multiple memebership ids. Then in the scope of this component, we can store the memberships ids that were created, and then in the undo, remove them in just one request by filtering the queryset using the `by_ids` filter.

AlexVelezLl reviewed (commented):
Hi @LianaHarris360! Just a couple of thoughts that could improve the performance here :).

LianaHarris360 committed:
Move data props and methods into setup() of EnrollLearnersSidePanel

LianaHarris360 committed:
Get facility user objects using by_ids

marcellamaki commented:
Thank you @pcenov - this is really helpful feedback. It seems like Liana will be tackling 3 and 4 in this PR, so some notes about the other items: 

> It seems that it should be further discussed whether the icons should be enabled when there are no users selected. For example if I have not selected a user and I click the "Enroll to class" icon I am just seeing a "You've selected 0 users" message and the "Enroll" button is disabled so there's nothing that I can actually do but go back, which brings me to my initial state

Yes, I agree. I think this should be resolved in https://github.com/learningequality/kolibri/pull/13513. When we tag the alpha release for you to test next week, let's confirm there (assuming both of these get merged) that this has been resolved as expected

> In continuation to the first point it's not immediately clear that the first icon with the back arrow is the "Assign coaches" button. The tooltip clears things out but to me this will always be a "Go back" icon.

As Jacob mentioned, I think having Tomiwa review this would be helpful 

3 & 4 will be resolved 

> Working on smaller screens with the Users table is quite difficult as it involves a lot of scrolling up and down and trying to find the appropriate options and controls - this can be seen in the video above.

Yes. This should be resolved in the course of my (very belated) PR to address "sticky columns" -- it first has to be merged into the design system and then updated in Kolibri in this table. I was hoping to have this in for the alpha, but we shall see if I make it 😅 

6, 7, & 8 - all very helpful notes. Point 8 I mentioned to Jacob and Liana in passing, but we will have to check in with Tomiwa. But yes, I think that having the opposite as the primary button makes the most sense. 

For 6 and 8, my inclination @nucleogenesis and @LianaHarris360 is that we can present an alternative suggestion to Tomiwa for a 👍 / 👎 over slack async and resolve at least those aspects quickly. For example, I think we can sort the classes alphabetically, and confirm that it would be okay to switch the primary and secondary buttons for the confirmation modal. I don't think they need to block this PR, but they could be relatively quick fixes and we could get them in for the alpha if Tomiwa has no objection. Then we can use the follow up conversation with Tomiwa to think through the more complex feedback that doesn't have as clear of a path (i.e. what icon should be used instead? how can we present the UI text information more concisely?, etc.)

LianaHarris360 committed:
Add by_ids filter to MembershipViewSet to filter by and bulk-delete memberships using a list of IDs

LianaHarris360 committed:
Keep track of newly created memberships and use by_ids filter to bulk-delete memberships

LianaHarris360 committed:
Fix undo enrollment header string

LianaHarris360 commented:
I've moved all of the computed properties and methods into `setup()`, updated the line-height of the `KModal` spans to remove the baby scroll, and implemented the performance updates Alex suggested (fetching `FacilityUsers`  and doing bulk membership deletion using `by_ids`). 

@pcenov I've corrected the string that previously stated "1 user have been enrolled" and for the enrolling a learner that was just deleted, we plan to create a follow-up issue to directly address that. Can you check if you're still receiving the 503 error "Database is not available for write operations" after enrolling several users to multiple classes and then clicking undo?

AlexVelezLl commented on kolibri/core/auth/api.py:
```diff
@@ -637,13 +637,17 @@ def get_queryset(self):
 
 class MembershipFilter(FilterSet):
     user_ids = CharFilter(method="filter_user_ids")
+    by_ids = CharFilter(method="filter_by_ids")
```
We also have an UUIDInFilter class we can use https://github.com/lianaharris360/kolibri/blob/187edf593d7c4088efaed8a2ccc64f97c06dc457/kolibri/core/auth/api.py#L319, but not sure would be the benefits of using it vs having a custom filter method.

LianaHarris360 replied:
The custom filter method allows us to filter by the id of the Membership model itself, so we can bulk delete the exact memberships we just created by passing in a list of ids.

AlexVelezLl commented on kolibri/core/auth/api.py:
```diff
@@ -652,7 +656,7 @@ class MembershipViewSet(BulkDeleteMixin, BulkCreateMixin, viewsets.ModelViewSet)
     serializer_class = MembershipSerializer
     filterset_class = MembershipFilter
-    filterset_fields = ["user", "collection", "user_ids"]
+    fields = ["user", "collection", "user_ids", "by_ids"]
```
I think the `filterset_fields` aren't needed anymore since we already have a filterset_class, or are they? However, was there a reason to change this to just `fields` here?

LianaHarris360 replied:
No reason, I had removed it initially because we already have the `filterset_class`, but decided to put it back because it was already not needed before my changes and I figured it was put there for a reason + I didn't want to change code that could potentially create a bug somewhere else in the codebase! Didn't mean to change it to `fields` instead of `filterset_fields`. I'll probably remove it and double check that it doesn't effect any other areas.

AlexVelezLl commented on kolibri/plugins/facility/assets/src/views/UserPage/SidePanels/EnrollLearnersSidePanel.vue:
```diff
@@ -1,40 +1,357 @@
+        classCoaches.value = [];
+        loading.value = true;
+        const membershipsPromise = MembershipResource.fetchCollection({
+          getParams: { user_ids: { user_ids: Array.from(props.selectedUsers).join(',') } },
```
Seems like we are loading all users here, since the user_ids filter is being incorrectly applied, its currently `getParams: { user_ids: { user_ids: ...} }` and it may just be `getParams: { user_ids: ... }`.

Here I have just 1 user selected, and the group by user is finding multiple users:

<img width="1947" height="318" alt="Image" src="https://github.com/user-attachments/assets/6d88a182-b545-42d5-b6ef-c0aca3c61923" />

LianaHarris360 replied:
Oops, thanks for catching this!

pcenov commented:
High @LianaHarris360 - I confirm that both the string and the error when undoing the enrollment of several users are fixed now.
A new issue which I've noticed is that when I try to enroll an already enrolled user, then the modal remains visible with a spinner icon and disabled buttons:

https://github.com/user-attachments/assets/21bf78f1-5717-4cb5-b4ff-0710fe178ded

AlexVelezLl reviewed (commented):
Seems like loading times have reduced significantly, this is looking great to me!

LianaHarris360 committed:
Correct usage of user_ids filter to only fetch memberships of selected users

LianaHarris360 commented:
@pcenov The loading value was not being reset after trying to enroll an already enrolled user, this should be fixed now.

AlexVelezLl reviewed (commented):
Nothing else to say on my side 😃. LGTM!

nucleogenesis reviewed (approved):
I just tested the last fix you added and it LGTM. Code looks good, so this is ready to ship! :shipit:

tomiwaoLE commented:
FYI latest set of design fixes: https://www.figma.com/design/2aE0zEXmyysxMZ6qO0ufh8?node-id=2734-4714#1359812771 @marcellamaki
